### ISSUE TITLE
Git/npm Cheatsheet

### ISSUE THREAD

**Issue by ethansaxenian** [ProfessorAmanda/econsimulations#136]
Here is a list of useful commands to know:

- `git branch` - displays a list of all branches in the project. The current branch will be marked with an asterix
- `git checkout branch_name` - switches to the branch named "branch_name"
- `git pull` - updates the local repository with the most recent changes on the current branch
- `npm install` - installs all dependencies described in `package.json` locally
- `npm start` - starts the development server

To view the site on `branch_name`:
```
git checkout branch_name
git pull
npm start
```

**If you get a** `module not found` **error**:
```
npm install
npm start
```

**Comment by tbyker** [comment]
Thank you!!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Ethan Saxenian ***@***.***>
Sent: Friday, July 16, 2021 1:03:37 PM
To: ProfessorAmanda/econsimulations ***@***.***>
Cc: Byker, Tanya S. ***@***.***>; Assign ***@***.***>
Subject: [ProfessorAmanda/econsimulations] Git/npm Cheatsheet (#136)


Here is a list of useful commands to know:

  *   git branch - displays a list of all branches in the project. The current branch will be marked with an asterix
  *   git checkout branch_name - switches to the branch named "branch_name"
  *   git pull - updates the local repository with the most recent changes on the current branch
  *   npm install - installs all dependencies described in package.json locally
  *   npm start - starts the development server

To view the site on branch_name:

git checkout branch_name
git pull
npm start


If you get a module not found error:

npm install
npm start

**Comment by ethansaxenian** [comment]
@tbyker if you want to checkout the 3d plot, it is one the branch named `3d-regression`. You will most likely need to run `npm install` before `npm start`

**Comment by ethansaxenian** [comment]
Beta distributions is on `master`.

You should still be able to run `git checkout 3d-regression`. Let me know if it doesn't work.

**Comment by tbyker** [comment]
Hi.  I am trying to look at the master and getting this error:
![image](https://user-images.githubusercontent.com/66968295/127510126-f23885b0-8e35-46ae-bd51-b51c269175c4.png)


I did:

git checkout master
git pull
npm install
npm start

**Comment by ethansaxenian** [comment]
Yikes, that's a bug, and it's also on the deployed website. I'll fix it and let's get it deployed quickly

**Comment by tbyker** [comment]
Maybe the bug affects all branches?  I was trying to look at the 3-d stuff (using the 3d-regression brand) and after spinning a long time I got this...  No rush if you are working on the other stuff.  
![image](https://user-images.githubusercontent.com/66968295/127516442-40267433-d274-4b28-9733-27b5f26618ac.png)
 
![image](https://user-images.githubusercontent.com/66968295/127516513-6cc76ae4-fc29-4f7d-92a2-698267c5df8d.png)

**Comment by ethansaxenian** [comment]
The bug should now be fixed on the master branch! Just run `git pull`

**Comment by ethansaxenian** [comment]
Oh just saw that second bug, no idea what that is...

**Comment by tbyker** [comment]
should i just try again?



**********************************************
Tanya Byker

Associate Professor, Economics

Middlebury College
***@***.******@***.***>
**********************************************



________________________________

**Comment by ProfessorAmanda** [comment]
For deleting old branches still stored locally: To delete a branch locally, run git branch -d [branchname]. Alternatively, I just did a quick google search and found this command: git branch --merged | egrep -v "(^\*|master|development)" | xargs git branch -d, which should delete all old local branches (source).

**Comment by ethansaxenian** [comment]
> For deleting old branches still stored locally: To delete a branch locally, run git branch -d [branchname]. Alternatively, I just did a quick google search and found this command: git branch --merged | egrep -v "(^*|master|development)" | xargs git branch -d, which should delete all old local branches (source).

Did the command work?

**Comment by ProfessorAmanda** [comment]
Yes, perfectly! Just thought I would store those commands here to close that other issue.

**Comment by ethansaxenian** [comment]
Great!

**Comment by ProfessorAmanda** [comment]
Some directions for the future, in case I merge a change from Dependabot that ends up breaking something. (Also directions here for pushing changes from my local).

If Dependabot install something you don't like after you merge a pull request, edit the package.json file in your local to roll back the version. Then do npm install in terminal. git status will show you what has been modified. To "transmit" these changes to the online repository, there are three steps: 1) Stage: 2) Commit 3) Push. To stage the changes, "git add ." Then check git status to see it green. To commit, type git commit -m 'whatever message you want'. Finally, git push.

**Comment by waynew99** [comment]
I’ve borrowed Ethan’s cheatsheet and made several changes – the most parts are still the same! I put [NEW] in front of the commands that are changed.

I’ve also added a deployment section for the deployment workflow.

## Useful commands
- `git branch` - displays a list of all branches in the project. The current branch will be marked with an asterix. Quit the view by hitting ‘q’.
- `git checkout branch_name` - switches to the branch named "branch_name"
- `git pull` - updates the local repository with the most recent changes on the **current** branch.
- [NEW]`npm install --legacy-peer-deps` - installs all dependencies described in `package.json` locally.
    - Due to the fact that one of our dependencies is not maintained properly, running `npm install` gives us an error. We need to append `--legacy-peer-deps` to any npm package management instructions. Thus, we should use `npm install --legacy-peer-deps`.
- [NEW]`npm run dev` - starts the development server*

To view the site on `branch_name`:
`git checkout branch_name`
`git pull`
[NEW]`npm run dev`
  - If you get a `module not found` error:
    - [NEW]`npm install --legacy-peer-deps`
    - [NEW]`npm run dev`

To delete all outdated local branches:
`git branch --merged | egrep -v "(^*|master|development)" | xargs git branch -d`


## Deployment process:
1. Make sure the master branch is ready to deploy – features work, style is correct, no linter errors, no test errors.
2. Create a pull request from `master` branch into `deployment` branch (the direction is important!)
- <img width="355" alt="Screen Shot 2022-06-29 at 12 11 53" src="https://user-images.githubusercontent.com/49931122/176485896-cd0c9544-5d65-4b65-af1e-4cf3fb5cb479.png">
3. Merge the pull request.
4. GitHub Action should do its job deploying everything. Monitor the Actions tab for any errors and address if needed.

**Comment by ProfessorAmanda** [comment]
Hi Wayne, I'm getting some errors when I try to run this locally. I started by running "git pull" for the master branch, then running "npm install --legacy-peer-deps". When I then tried to run "npm run dev," I error that said something like "dev not found." Then I just ran it using "npm start," and was able to run it locally. However, then when I ran "git pull" to make sure everything was up to date, I got an error "error: Your local changes to the following files would be overwritten by merge:
	package-lock.json
Please commit your changes or stash them before you merge." Can you recommend a next step?

**Comment by waynew99** [comment]
Hi Amanda,
It seems like react changed something in package-lock.json and that prevents a git pull. Let's try `git stash` before git pull to get rid of local changes. So please try:
```
git stash
git pull
npm install --legacy-peer-deps
npm start dev
````

Hope it works!

**Comment by ProfessorAmanda** [comment]
Okay, I was able to run it! But when I try to check out another branch, I get an error: "error: Your local changes to the following files would be overwritten by checkout:
	package-lock.json
Please commit your changes or stash them before you switch branches.
Aborting."

I'm going to stash the changes for now, and we can maybe chat during our meeting.

**Comment by ProfessorAmanda** [comment]
Hey Wayne, I got this error again just now when I tried to check out NormalDistribution: "Midd-19529:econsimulations agregg$ git checkout NormalDistribution
error: Your local changes to the following files would be overwritten by checkout:
	package-lock.json"

Do you think there might be a tweak we need to make to package-lock.json to make everything agree? There might just be a version number or something that differs among our versions of the repo.

For now I've just stashed the changes again to see what's up with NormalDistributions.

**Comment by waynew99** [comment]
Thanks for the detailed error message. I did a little research and it seems like that there's a difference between the node version of your computer's environment and the node version of our project. I'll try finding potential solutions for us to try out next time we meet. For now, stashing them should be ok.

**Comment by ProfessorAmanda** [comment]
Btw my issues with package-lock.json seem to have resolved? At least, I'm not having any problems or needing to stash changes when I work with the NormalDistribution branch.

**Comment by waynew99** [comment]
I'm not sure what the root cause was, but I'm glad it's not there anymore! If it still occurs with other branches, we can try several solutions next time when we meet.
